### PR TITLE
Remove `logprob_grad_fn` kwargs in favor of `jax.custom_vjp`

### LINKED ISSUE #341: Remove the `logprob_grad` keyword arguments
To use custom gradients users currently need to pass the function as a keyword arguments; this makes the API more complicated as we need to carry this argument downstream.

We can bypass this thanks to JAX allowing to register a custom gradient for functions. This is related to #318 and also #340 and probably #319 as well.

### PR DESCRIPTION
**Pull request by rlouf**
Custom gradient currently need to be passed to the algorithms all the way down to the integrators. However JAX has a much nicer way of doing that; you can write a logprob and register a gradient for that logprob; JAX will then use it when `jax.grad` is called on that function.

Closes #341 #318 

 - [x] We should be able to understand what the PR does from its title only;
 - [x] There is a high-level description of the changes;
 - [x] There are links to *all* the relevant issues, discussions and PRs;
 - [x] The branch is rebased on the latest `main` commit;
 - [x] Commit messages follow these [guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html);
 - [x] The code respects the current naming conventions;
 - [x] Docstrings follow the [numpy style guide](https://numpydoc.readthedocs.io/en/latest/format.html)
 - [x] `pre-commit` is installed and configured on your machine, and you ran it before opening the PR;
 - [x] There are tests covering the changes;
 - [x] The doc is up-to-date;

### CODE REVIEW

**Comment by rlouf** [comment]
We still need a small regression test to make sure that the new solution (relying on JAX's `custom_jvp` or `custom_vjp`) works, something with the `velocity_verlet` integrator should do, and to add a good example that illustrates the need for custom gradients (functions that contain `minimize` for instance).

**Comment by rlouf** [comment]
One thing I've noticed: MEADS compute the gradients during the warmup; however those are computed by the GHMC step. We should use that instead.

File: blackjax/adaptation/meads.py [hunk 1]
```diff
@@ -26,10 +26,7 @@ class MEADSAdaptationState(NamedTuple):
     delta: float
 
 
-def base(
-    logprob_grad_fn: Callable,
-    batch_fn: Callable = jax.vmap,
-):
+def base():
```
Comment (junpenglao) [inline blackjax/adaptation/meads.py:29]: Seems unnecessary - should we just remove this and let the function exists in the namespace?

File: examples/howto_custom_gradients.md [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+  name: python3
+---
+
+# Use custom gradients
+
+
+JAX provides a convenient `jax.grad` function to evaluate the gradient of any function build with JAX primitives. Which is why Blackjax uses `jax.grad` internally whenever it needs to evaluate the gradient. This should be enough for most applications, but sometimes you may need to provide your own gradients to blackjax for several reasons:
+
+- You have a convenient closed-form expression for the gradient that is evaluated faster than the gradient that JAX produces;
+- The forward-mode differentiation is faster than the backward-mode;
```
Comment (junpenglao) [inline examples/howto_custom_gradients.md:20]: This is actually the initial motivation of adding this API - could you add a small example of `custom_jvp` with `jax.jacfwd()`?

Comment (rlouf) [reply examples/howto_custom_gradients.md:20]: Do you have an example of a logprob where it makes sense to use `jax.jaxfwd()` instead of `jax.grad`?

Comment (junpenglao) [reply examples/howto_custom_gradients.md:20]: [FedericoV](https://github.com/FedericoV) mentioned a stiff ODE: https://github.com/blackjax-devs/blackjax/issues/201#issuecomment-1255592363

Comment (rlouf) [reply examples/howto_custom_gradients.md:20]: I didn't see how to do it immediately; I think the best way is to do something like

```python
import jax.numpy as jnp

def grad(x):
    return jnp.dot(x, jax.jacfwd(logprob_fn)(x))

# Override `jax.grad` in the namespace
```
